plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/TestBridgingHiveMetastore.java

```diff
@@ -47,4 +68,97 @@ void afterAll()
+
+        metastore.dropDatabase(databaseName, false);
+    }
+
+    @Test
+    public void testCreateDatabaseFailsAfter10Retries()
+    {
+        // Simulate 10 timeouts
+        Map<Operation, Queue<Modification>> metastoreClientBehaviourModifiers = Map.of(Operation.CREATE_DATABASE,
+                new ArrayDeque<>(Collections.nCopies(10,
+                        new Modification(Type.INSTEAD, () -> {
+                            throw new RuntimeException("timeout");
+                        }))));
+
+        modifyMetastoreClientBehaviour(metastoreClientBehaviourModifiers);
+
+        String databaseName = "test_database_" + randomNameSuffix();
+        Database.Builder database = Database.builder()
+                .setDatabaseName(databaseName)
+                .setOwnerName(Optional.empty())
+                .setOwnerType(Optional.empty());
+
+        // Test timeout failure
+        assertThatThrownBy(() -> metastore.createDatabase(database.build()))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessage("timeout");
+    }
+
+    @Test
+    public void testCreateTableWithRetries()
+    {
+        // Simulate timeout after first create table
+        Map<Operation, Queue<Modification>> metastoreClientBehaviourModifiers = Map.of(Operation.CREATE_TABLE, new ArrayDeque<>(List.of(
+                new Modification(Type.AFTER, () -> {
+                    throw new RuntimeException();
+                }))));
+        modifyMetastoreClientBehaviour(metastoreClientBehaviourModifiers);
+
+        String databaseName = "test_database_" + randomNameSuffix();
+        Database.Builder database = Database.builder()
+                .setDatabaseName(databaseName)
+                .setOwnerName(Optional.empty())
+                .setOwnerType(Optional.empty());
+        metastore.createDatabase(database.build());
+
+        String tableName = "test_table" + randomNameSuffix();
+        Table.Builder table = Table.builder()
+                .setDatabaseName(databaseName)
+                .setTableName(tableName)
+                .setParameters(Map.of(TRINO_QUERY_ID_NAME, "query_id"))
+                .setTableType(EXTERNAL_TABLE.name())
+                .setOwner(Optional.empty());
+        table.getStorageBuilder()
+                .setStorageFormat(fromHiveStorageFormat(PARQUET));
+        metastore.createTable(table.build(), NO_PRIVILEGES);
+
+        table.setParameters(Map.of(TRINO_QUERY_ID_NAME, "another_query_id"));
+        assertThatThrownBy(() -> metastore.createTable(table.build(), NO_PRIVILEGES))
+                .isInstanceOf(TableAlreadyExistsException.class);
+
+        metastore.dropTable(databaseName, tableName, false);
+        metastore.dropDatabase(databaseName, false);
+    }
+
+    private void modifyMetastoreClientBehaviour(Map<Operation, Queue<Modification>> behaviourModifiers)
+    {
+        setMetastore(new BridgingHiveMetastore(testingThriftHiveMetastoreBuilder()
```

Comment:
`setMetastore` seems to have been thought as a method to be called only from the constructor.

Side-note: It would probably make more sense to have a `getMetastore()` method in the abstract class and initiate the `metastore` as a `private final` field in the concrete classes.